frontend/src/metabase/query_builder/components/dataref/TablePane.jsx

```diff
@@ -1,85 +1,85 @@
 /* eslint "react/prop-types": "warn" */
-import React, { Component } from "react";
+import React from "react";
 import PropTypes from "prop-types";
+import { connect } from "react-redux";
 import { t } from "ttag";
 import cx from "classnames";
-import Icon from "metabase/components/Icon";
 
 // components
+import Icon from "metabase/components/Icon";
 import Expandable from "metabase/components/Expandable";
 
 // lib
-import { createCard } from "metabase/lib/card";
-import * as Q_DEPRECATED from "metabase/lib/query";
 import { foreignKeyCountsByOriginTable } from "metabase/lib/schema_metadata";
 import { inflect } from "metabase/lib/formatting";
 
-export default class TablePane extends Component {
-  constructor(props, context) {
-    super(props, context);
-    this.setQueryAllRows = this.setQueryAllRows.bind(this);
-    this.showPane = this.showPane.bind(this);
-
-    this.state = {
-      table: undefined,
-      tableForeignKeys: undefined,
-      pane: "fields",
-    };
-  }
+// entities
+import Table from "metabase/entities/tables";
+
+const mapStateToProps = (state, ownProps) => ({
+  tableId: ownProps.table.id,
+  table: Table.selectors.getObject(state, { entityId: ownProps.table.id }),
+});
+
+const mapDispatchToProps = {
+  fetchForeignKeys: Table.actions.fetchForeignKeys,
+  fetchMetadata: Table.actions.fetchMetadata,
+};
+
+@connect(
+  mapStateToProps,
+  mapDispatchToProps,
+)
+export default class TablePane extends React.Component {
+  state = {
+    pane: "fields",
+    error: null,
+  };
 
   static propTypes = {
     query: PropTypes.object.isRequired,
-    loadTableAndForeignKeysFn: PropTypes.func.isRequired,
     show: PropTypes.func.isRequired,
     onClose: PropTypes.func.isRequired,
     setCardAndRun: PropTypes.func.isRequired,
+    tableId: PropTypes.number.isRequired,
     table: PropTypes.object,
+    fetchForeignKeys: PropTypes.func.isRequired,
+    fetchMetadata: PropTypes.func.isRequired,
   };
 
-  componentWillMount() {
-    this.props
-      .loadTableAndForeignKeysFn(this.props.table.id)
-      .then(result => {
-        this.setState({
-          table: result.table,
-          tableForeignKeys: result.foreignKeys,
-        });
-      })
-      .catch(error => {
-        this.setState({
-          error: t`An error occurred loading the table`,
-        });
+  async componentWillMount() {
+    try {
+      await Promise.all([
+        this.props.fetchForeignKeys({ id: this.props.tableId }),
+        this.props.fetchMetadata({ id: this.props.tableId }),
+      ]);
```

Comment:
Not worth doing here, but what do you think about having object loaders take `objectActionLoaders: ['fetchMetadata']` or something that waits on object actions before removing the loading spinner?

Reply:
Something like that might make sense, but currently we use the `getLoading` selector which I think is more robust than waiting on the promise that's returned. I'll see what I can come up with.